x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/InternalMapReduceAggregation.java

```diff
@@ -0,0 +1,126 @@
+import org.elasticsearch.search.aggregations.AggregationReduceContext;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.support.SamplingContext;
+import org.elasticsearch.search.profile.SearchProfileResults;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public final class InternalMapReduceAggregation extends InternalAggregation {
+
+    private final AbstractMapReducer<?, ?, ?> mapReducer;
+    private Writeable mapReduceContext;
+    private final boolean profiling;
+
+    InternalMapReduceAggregation(
+        String name,
+        Map<String, Object> metadata,
+        AbstractMapReducer<?, ?, ?> mapReducer,
+        Writeable mapReduceContext,
+        boolean profiling
+    ) {
+        super(name, metadata);
+        this.mapReducer = Objects.requireNonNull(mapReducer);
+        this.mapReduceContext = Objects.requireNonNull(mapReduceContext);
+        this.profiling = profiling;
+    }
+
+    public InternalMapReduceAggregation(StreamInput in) throws IOException {
+        super(in);
+        this.mapReducer = in.readNamedWriteable(AbstractMapReducer.class);
+
+        // TODO: we should not use `NON_RECYCLING_INSTANCE` here, however we don't have access to AggregationContext yet
+        this.mapReduceContext = this.mapReducer.readMapContext(in, BigArrays.NON_RECYCLING_INSTANCE);
+        this.profiling = in.readBoolean();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return mapReducer.getAggregationName();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+
+        out.writeNamedWriteable(mapReducer);
+        mapReducer.writeContext(out, mapReduceContext);
+        out.writeBoolean(profiling);
+    }
+
+    @Override
+    public InternalAggregation reduce(List<InternalAggregation> aggregations, AggregationReduceContext reduceContext) {
+        // TODO: as long as we haven't clarified who is closing, we can't use big arrays from the context
+        Writeable newMapReduceContext = mapReducer.reduceInit(BigArrays.NON_RECYCLING_INSTANCE);
+        // MapReduceContext newMapReduceContext = mapReducer.reduceInit(reduceContext.bigArrays());
+
+        // TODO: if we use the recycling big array instance and something fails here the internal aggregations aren't closed
+        if (reduceContext.isFinalReduce()) {
+            mapReducer.reduce(
+                aggregations.stream().map(agg -> ((InternalMapReduceAggregation) agg).getMapReduceContext()),
+                newMapReduceContext
+            );
+            try {
+                newMapReduceContext = mapReducer.reduceFinalize(newMapReduceContext);
+            } catch (IOException e) {
+                throw new AggregationExecutionException("Final reduction failed", e);
+            }
```

Comment:
I still think that if we are to use this `mapReducer` framework, types need to be better defined. Right now `doReduceFinalize` simply takes a writeable and returns a writeable (might as well be an opaque object) and the eclat reducer switches types. Now `newMapReduceContext` gets passed to `reduceFinalize` as a `TransactionStore` and then the type complete changes to `EclatResult`. 

For example, `TransactionStore` doesn't need to be written to XContent. But `EclatResult` does.

The map reducer already has `MapReduceContext` vs a `MapReduceResult` why not use those types higher up the call chain instead of casting?